### PR TITLE
Implement learning rate finder

### PR DESCRIPTION
**Pull request by mjacar**
# What is this?
This is an implementation of a learning rate finder as implemented in [the fastai library](https://github.com/fastai/fastai/blob/79ecfd5c60087766c6f1952e674f98470d488069/fastai/callbacks/one_cycle.py) and as initially introduced in the paper [Super-Convergence: Very Fast Training of Neural Networks Using Large Learning Rates](https://arxiv.org/abs/1506.01186).
# Why is this important?
The process of finding an optimal learning rate is generally a highly expensive process that involves running numerous identical experiments with different learning rates (determined either via grid search or random search) and simply observing which learning rate converges fastest. Instead, as introduced in the paper [Super-Convergence: Very Fast Training of Neural Networks Using Large Learning Rates](https://arxiv.org/abs/1506.01186), we can take a much more efficient approach. In the LR range test, we do a simulated training run of our model where we increase the learning rate at each step. We then get a plot of training loss vs. learning rate. We can then see the highest learning rate our model will admit before it starts to diverge. It has been shown that using a learning rate (or max learning rate if you're using a learning rate scheduler) that corresponds to the point of the greatest negative rate of change is a good choice. This entire process takes minutes at most and doesn't require running any experiments in parallel.
# How was this tested?
I wasn't really sure what unit tests to add, but I did make a gist [demonstrating how it works](https://gist.github.com/mjacar/0ab81ef03f42d85c4716e0e1c4240ab9).
# Other Comments
1. I don't think there's another utility in pytorch that is similar to this, so I'm aware this is a fairly speculative proposal. Nevertheless, I feel like something like this ought to exist in pytorch.
1. Given that, I wasn't 100% positive where the best place to put this was. If someone has some better idea, then I'm all ears.
1. This requires `matplotlib` in order to work. I don't know how people feel about that, but it's worth pointing out. I don't think you can plot arbitrary line graphs on Tensorboard.
1. I also wasn't sure how to actually add the documentation.

### CODE REVIEW

**Review by soumith — CHANGES_REQUESTED** [review]
this is too speculative of a method to really belong to pytorch core.
I dont think I'd be inclined to accept it.

separately, tensorboard is not a great place to put it, it doesn't fit with anything wrt the tensorboard integration
